File: typescript/day3_part1_2017.ts
import { readFileSync } from 'fs';

const calculateSteps = (input: number): number => {
    if (input === 1) return 0;

    let layer = Math.ceil((Math.sqrt(input) - 1) / 2);
    let maxInLayer = (2 * layer + 1) ** 2;
    let sideLength = 2 * layer;
    let offset = (maxInLayer - input) % sideLength;

    let distanceToCenter = Math.abs(offset - layer);
    return layer + distanceToCenter;
};

const main = () => {
    const data = readFileSync('input.txt', 'utf-8').trim();
    const input = parseInt(data, 10);
    const steps = calculateSteps(input);
    console.log(steps);
};

main();

File: clojure/day3_part1_2017.clj
(ns solution
  (:import (java.io BufferedReader FileReader))
  (:import (java.lang Math)))

(defn main []
  (with-open [r (BufferedReader. (FileReader. "input.txt"))]
    (let [target (Integer/parseInt (.trim (first (line-seq r))))]
      (let [side-length (int (Math/ceil (Math/sqrt target)))]
        (if (even? side-length) (inc side-length) side-length)
        (let [max-value (* side-length side-length)
              steps-from-edge (quot (dec side-length) 2)
              distance-to-middle (apply min (for [i (range 4)]
                                             (let [middle-point (- max-value steps-from-edge (* (dec side-length) i))]
                                               (Math/abs (- target middle-point)))))]
          (println (+ steps-from-edge distance-to-middle)))))))

(main)